lib/compat/wordpress-6.1/class-gutenberg-rest-templates-controller.php

```diff
@@ -10,6 +10,54 @@
  * Base Templates REST API Controller.
  */
 class Gutenberg_REST_Templates_Controller extends WP_REST_Templates_Controller {
+
+	/**
+	 * Registers the controllers routes.
+	 *
+	 * @return void
+	 */
+	public function register_routes() {
+		// Get fallback template content.
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base . '/fallback_content',
+			array(
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_template_fallback_content' ),
+					'permission_callback' => array( $this, 'get_items_permissions_check' ),
+					'args'                => array(
+						'slug' => array(
+							'description' => __( 'The slug of the template to get the fallback content for', 'gutenberg' ),
+							'type'        => 'string',
+						),
+					),
+				),
+			)
+		);
+		parent::register_routes();
+	}
+
+	/**
+	 * Returns the fallback template content for a given slug.
+	 *
+	 * @param WP_REST_Request $request The request instance.
+	 *
+	 * @return WP_REST_Response|WP_Error
+	 */
+	public function get_template_fallback_content( $request ) {
+		if ( empty( $request['slug'] ) ) {
+			return new WP_Error(
+				'rest_invalid_param',
+				__( 'Invalid slug.', 'gutenberg' ),
+				array( 'status' => 400 )
+			);
+		}
+		$template_hierarchy = gutenberg_get_template_hierarchy( $request['slug'] );
+		$fallback_template  = resolve_block_template( $request['slug'], $template_hierarchy, '' );
```

Comment:
So per [my other comment](https://github.com/WordPress/gutenberg/pull/42520/files#r924312884), here is where I'd loop over template types and only create `$template_hierarchy` arrays for the current template type (basically mimicking [this loop](https://github.com/WordPress/wordpress-develop/blob/9db156e7c73774836e3da71d8aac8c0adc2cf5e9/src/wp-includes/template-loader.php#L78-L95)).

Something like

```php
$template_types = array(
	'embed' => 'gutenberg_get_embed_template_hierarchy',
	'404' => 'gutenberg_get_404_template_hierarchy',
	'search' => 'gutenberg_get_search_template_hierarchy',
	'front-page' => 'gutenberg_get_front_page_template_hierarchy',
	'home' => 'gutenberg_get_home_template_hierarchy',
	/* ... */
);

foreach ( $template_types as $template_type => $template_hierarchy_getter ) {
	// Fortunately, we don't need any sophisticated `is_${template_type}()` checks here -- a simple "starts_with" is enough.
	if ( strpos( $request['slug'], $template_type  ) === 0 ) {
		$template_hierarchy = call_user_func( $template_hierarchy_getter, $request['slug'] );
		$fallback_template = resolve_block_template( $request['slug'], $template_hierarchy, '' );
		if ( $fallback_template ) {
			break;
		}
	}

	// We might want to absorb the following into `$template_types`.
	if ( ! $fallback_template ) {
		$fallback_template = resolve_block_template( $request['slug'], array( 'index' ), '' );
	}
}
```

Reply:
(This should also allow for better unit testing of the `gutenberg_get_${template_type}_template_hierarchy` functions.)